fsx/main.tf

```diff
@@ -0,0 +1,327 @@
+  subnet_id      = aws_subnet.public.id
+}
+
+# Umbrella security group for Batch compute environments & filesystem mount targets, allowing any
+# traffic within the VPC and outbound-only Internet access.
+# The ingress could be locked down to allow only FSxL traffic within the VPC,
+# https://docs.aws.amazon.com/fsx/latest/LustreGuide/limit-access-security-groups.html#fsx-vpc-security-groups
+resource "aws_security_group" "all" {
+  name   = var.environment_tag
+  vpc_id = aws_vpc.vpc.id
+  ingress {
+    from_port   = 0
+    to_port     = 0
+    protocol    = "-1"
+    cidr_blocks = [aws_vpc.vpc.cidr_block]
+  }
+  # Uncomment to open SSH access via EC2 Instance Connect
+  /*
+  ingress {
+    from_port   = 22
+    to_port     = 22
+    protocol    = "tcp"
+    cidr_blocks = ["0.0.0.0/0"]
+  }
+  */
+  egress {
+    from_port   = 0
+    to_port     = 0
+    protocol    = "-1"
+    cidr_blocks = ["0.0.0.0/0"]
+  }
+}
+
+/**************************************************************************************************
+ * FSxL
+ *************************************************************************************************/
+
+resource "aws_fsx_lustre_file_system" "lustre" {
+  subnet_ids                    = [aws_subnet.public.id]
+  security_group_ids            = [aws_security_group.all.id]
+  deployment_type               = "SCRATCH_2"
+  storage_capacity              = var.lustre_GiB
+  weekly_maintenance_start_time = var.lustre_weekly_maintenance_start_time
+
+  lifecycle {
+    prevent_destroy = true
+  }
+}
+
+/**************************************************************************************************
+ * Batch
+ *************************************************************************************************/
+
+data "cloudinit_config" "all" {
+  # user data scripts for Batch worker instances
+  gzip = false
+
+  part {
+    content_type = "text/x-shellscript"
+    content      = file("${path.module}/../assets/init_docker_instance_storage.sh")
+  }
+
+  part {
+    content_type = "text/x-shellscript"
+    content      = <<-EOT
+    #!/bin/bash
+    set -euxo pipefail
+    # enable EC2 Instance Connect for troubleshooting (if security group allows inbound SSH)
+    yum install -y ec2-instance-connect && grep eic_run_authorized_keys /etc/ssh/sshd_config
+    # mount FSxL
+    amazon-linux-extras install -y lustre2.10
+    mkdir -p /mnt/fsx
+    mount -t lustre -o noatime,flock ${aws_fsx_lustre_file_system.lustre.dns_name}@tcp:/${aws_fsx_lustre_file_system.lustre.mount_name} /mnt/fsx
+    lfs setstripe -E 1G -c 1 -E 16G -c 4 -S 16M -E -1 -c -1 -S 256M /mnt/fsx
+    df -h
+    # Somehow the preceding steps nondeterministically interfere with ECS agent startup. Set a cron
+    # job to keep trying to start it. (We can't simply `systemctl start ecs` here, because the ecs
+    # systemd service requires cloud-init to have finished.)
+    echo "* * * * * root /usr/bin/systemctl start ecs" > /etc/cron.d/ecs-workaround
```

Comment:
@markjschreiber @wleepang I got stuck for awhile on this really weird problem where the lustre install/mount procedure somehow prevented the ECS agent from starting up on the batch worker instances (using the default AMI). No error messages nor even any log file in /var/log/ecs, it seemingly just never even tries to start. Have you heard about anything like this? Workaround here that seems effective.

Reply:
I know that in Lee's AWS genomics repo the ECS agent is stopped before the script runs and then restarted after although I assume this is because we move the docker setup to a different volume?

Reply:
Correct, the ecs-agent stop and restart is because of configuration changes to docker. Not sure if a restart is currently needed for fsx. It wasn't an issue in the past.